### DISCUSSION TITLE
Hooking an Encrypted Game?

### DISCUSSION THREAD

**Discussion by joshfreitas1984** [BepInEx/BepInEx#289] [Help and Q&A]
Hi guys,

Is it possible to hook with BepInEx an encrypted game and still use Harmony after the game has loaded the decrypted DLL into memory?

I am currently translating a game at the moment with an encrypted Assembly-CSharp.dll. I have a decrypted version but obviously it wont load with the game because the game still tries to decrypt an already decrypted dll. I currently use cheat engine and AOBScans to find some of the addresses I want to change because once the game is loaded, the full assembly is decrypted into memory.

Is it safe to assume that without removing the decrypt method from the executable, I wont be able to use harmony to patch out the methods I want to override? Or will it still be possible because the assembly is decrypted in memory?

Right now I'm unable to hook the game at all with BepIn which I am assuming is because of the encryption. Can I even use this at all or will I have to investigate how to manually inject through mono?

Any direction you could point me in would be great!

Thanks

Josh

**Comment by ghorsington** [comment]
This depends mostly on the game and how it handles encryption. If the decryption code is baked into the UnityPlayer.dll or the game executable, you could try getting stock executable and engine by downloaded the correct Unity Editor version from [Unity Download Archive](https://unity3d.com/get-unity/download/archive). The stock player and engine are usually found in `<unity-install-dir>\Editor\Data\PlaybackEngines\windowsstandalonesupport\Variations`.
If you're lucky and there is no additional special code besides assembly encryption, with stock engine binaries you should be able to boot the game using decrypted DLLs.

Does BepInEx itself bootstrap fine? Are you able to get BepInEx console and have plugins loaded?
If so, given that BepInEx itself is bootstrapped even before Assembly-CSharp.dll is loaded from the file, I'd actually try making a preloader patch in which you can try to load the decrypted Assembly-CSharp.dll via `Assembly.Load` directly. If that works, the game won't be able to load the encrypted version as there is already one loaded into mono's AppDomain.

Finally, if you don't want to modify anything, you should be able to use the general Reflection API like `AppDomain.Current.GetAssemblies()` to get the assembly and the types you want to patch. Harmony supports manual patching via [`Harmony.Patch()`](https://harmony.pardeike.net/api/HarmonyLib.Harmony.html#HarmonyLib_Harmony_Patch_System_Reflection_MethodBase_HarmonyLib_HarmonyMethod_HarmonyLib_HarmonyMethod_HarmonyLib_HarmonyMethod_HarmonyLib_HarmonyMethod_) so you could patch anything in an assembly without having to reference it at all in your code. It's a bit of a mess but can be used.

**Reply by joshfreitas1984** [reply]
BepInEx doesnt get to the console at all at the moment just crashes straight away.

However I'll give the pre-loader a go and the base stock executables a shot :) If I can get it loading with either I should be ok from there!
